<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#План-исследования" data-toc-modified-id="План-исследования-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>План исследования</a></span></li></ul></li><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Предобработка данных</a></span></li></ul></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Обучающая-выборка" data-toc-modified-id="Обучающая-выборка-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обучающая выборка</a></span></li><li><span><a href="#Валидационная-выборка-и-тестовая-выборки" data-toc-modified-id="Валидационная-выборка-и-тестовая-выборки-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Валидационная выборка и тестовая выборки</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Исследуем-модель-Decision-Tree-Classifier" data-toc-modified-id="Исследуем-модель-Decision-Tree-Classifier-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Исследуем модель Decision Tree Classifier</a></span></li><li><span><a href="#Исследуем-модель-Random-Forest-Classifier" data-toc-modified-id="Исследуем-модель-Random-Forest-Classifier-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Исследуем модель Random Forest Classifier</a></span></li><li><span><a href="#Исследуем-модель-Logistic-Regression" data-toc-modified-id="Исследуем-модель-Logistic-Regression-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Исследуем модель Logistic Regression</a></span></li><li><span><a href="#Вывод-по-разделу-исследования-моделей" data-toc-modified-id="Вывод-по-разделу-исследования-моделей-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span><em>Вывод по разделу исследования моделей</em></a></span></li></ul></li><li><span><a href="#Проверка-моделей-на-тестовой-выборке" data-toc-modified-id="Проверка-моделей-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка моделей на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Проверка-Decision-Tree-Classifier-на-тестовой-выборке" data-toc-modified-id="Проверка-Decision-Tree-Classifier-на-тестовой-выборке-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Проверка Decision Tree Classifier на тестовой выборке</a></span></li><li><span><a href="#Проверка-Random-Forest-Classifier-на-тестовой-выборке" data-toc-modified-id="Проверка-Random-Forest-Classifier-на-тестовой-выборке-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Проверка Random Forest Classifier на тестовой выборке</a></span></li><li><span><a href="#Проверка-Logistic-Regression-на-тестовой-выборке" data-toc-modified-id="Проверка-Logistic-Regression-на-тестовой-выборке-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Проверка Logistic Regression на тестовой выборке</a></span></li><li><span><a href="#Вывод-по-разделу-проверки-моделей-на-тестовой-выборке" data-toc-modified-id="Вывод-по-разделу-проверки-моделей-на-тестовой-выборке-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span><em>Вывод по разделу проверки моделей на тестовой выборке</em></a></span></li></ul></li><li><span><a href="#Проверка-модели-на-адекватность" data-toc-modified-id="Проверка-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка модели на адекватность</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><em>Общий вывод</em></a></span></li></ul></div>

# Рекомендация тарифов

### Описание проекта

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

### План исследования

1. Обзор и предобработка данных: 
- Посмотреть общую информацию о данных
- Выявить и устранить дупликаты и пропуски (если это необходимо)
- Изменить тип данных при необходимости

2. Разбить данные на выборки
- Обучающую, валидационную и тестовую

3. Исследовать модели
- Decision Tree Classifier
- Random Forest Classifier
- Logistic Regression

4. Проверить модели на тестовой выборке
5. Проверить модель на адекватность
6. Общий вывод

## Откройте и изучите файл

### Предобработка данных

План этапа: 
- Посмотреть общую информацию о данных
- Выявить и устранить дупликаты и пропуски (если это необходимо)
- Изменить тип данных при необходимости

In [1]:

import pandas as pd
import optuna 
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score

In [3]:
try: 
    data = pd.read_csv('users_behavior.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv')

In [47]:
data.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [48]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [49]:

for col in data[['calls','messages']]:
    data[col] = data[col].astype(int)

In [50]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   int64  
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   int64  
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 125.7 KB


In [51]:
data['is_ultra'].unique()

array([0, 1])

In [52]:

col_list = data.columns
col_list

Index(['calls', 'minutes', 'messages', 'mb_used', 'is_ultra'], dtype='object')

In [53]:

print('кол-во дупликатов:', data.duplicated().sum())
for col in col_list:
    print('-'* 25)
    print(col,': кол-во NaN',data[col].isna().sum(),
          ', процент NaN', round(data[col].isna().sum()/len(data)*100, 2),'%')

кол-во дупликатов: 0
-------------------------
calls : кол-во NaN 0 , процент NaN 0.0 %
-------------------------
minutes : кол-во NaN 0 , процент NaN 0.0 %
-------------------------
messages : кол-во NaN 0 , процент NaN 0.0 %
-------------------------
mb_used : кол-во NaN 0 , процент NaN 0.0 %
-------------------------
is_ultra : кол-во NaN 0 , процент NaN 0.0 %


Был изменен тип столбцов calls и messages на int для удобства исследования, в остальном замечаний к данным нет

## Разбейте данные на выборки

План этапа: 
- Разбить данные на выборки:
  - Обучающую
  - Валидационную и тестовую

Данные разбиваем на три части: обучающую, валидационную и тестовую - 3:1:1

### Обучающая выборка

In [54]:

train_data, valid_data = train_test_split(data, test_size = 0.25, random_state = 12345)
print(f'Размер обучающей выборки: {train_data.shape}')


Размер обучающей выборки: (2410, 5)


### Валидационная выборка и тестовая выборки

In [55]:

valid_data, final_test_data = train_test_split(valid_data, test_size = 0.5, random_state = 12345)
print(f'Размер валидационной выборки: {valid_data.shape}')
print(f'Размер тестовой выборки: {final_test_data.shape}')


Размер валидационной выборки: (402, 5)
Размер тестовой выборки: (402, 5)


## Исследуйте модели

План этапа:  
Исследовать модели:  
- Decision Tree Classifier
- Random Forest Classifier
- Logistic Regression

### Исследуем модель Decision Tree Classifier

In [56]:
# Подготовка данных для исследования

train_features = data.drop(['is_ultra'], axis = 1)
train_target = data['is_ultra']

valid_features = valid_data.drop(['is_ultra'], axis = 1)
valid_target = valid_data['is_ultra']


In [57]:
# Построение цикла с обучением и предсказанием модели

best_result_dtc = 0

for depth in range(3, 11):
    model_dtc = DecisionTreeClassifier(random_state = 12345, max_depth = depth)
    model_dtc.fit(train_features, train_target)
    
    predictions_train_dtc = model_dtc.predict(train_features)
    predictions_valid_dtc = model_dtc.predict(valid_features)
    
    score_train_dtc = accuracy_score(predictions_train_dtc, train_target)
    score_valid_dtc = accuracy_score(predictions_valid_dtc,valid_target)
    
    print(f'Accuracy при max_depth = {depth}')
    print(f'Обучающая выборка: {score_train_dtc}')
    print(f'Валидационная выборка: {score_valid_dtc}')
    print()
    
    if score_valid_dtc > best_result_dtc:
        best_result_dtc = score_valid_dtc
        best_depth_dtc = depth
        
    
print(f'Лучший показатель Accuracy: {best_result_dtc.round(4)} при depth = {best_depth_dtc}')

Accuracy при max_depth = 3
Обучающая выборка: 0.798070939639079
Валидационная выборка: 0.8034825870646766

Accuracy при max_depth = 4
Обучающая выборка: 0.7999377722464219
Валидационная выборка: 0.8109452736318408

Accuracy при max_depth = 5
Обучающая выборка: 0.8170504044803982
Валидационная выборка: 0.8084577114427861

Accuracy при max_depth = 6
Обучающая выборка: 0.827317983820784
Валидационная выборка: 0.818407960199005

Accuracy при max_depth = 7
Обучающая выборка: 0.8397635345364033
Валидационная выборка: 0.8233830845771144

Accuracy при max_depth = 8
Обучающая выборка: 0.8550093341630367
Валидационная выборка: 0.8383084577114428

Accuracy при max_depth = 9
Обучающая выборка: 0.8637212196639701
Валидационная выборка: 0.8557213930348259

Accuracy при max_depth = 10
Обучающая выборка: 0.8721219663970131
Валидационная выборка: 0.8706467661691543

Лучший показатель Accuracy: 0.8706 при depth = 10


### Исследуем модель Random Forest Classifier

In [58]:
# Построение цикла с обучением и предсказанием модели

best_result_rfc = 0

for est in range(3, 11):
    model_rfc = RandomForestClassifier(random_state = 12345, n_estimators = est)
    model_rfc.fit(train_features, train_target)
    
    predictions_train_rfc = model_rfc.predict(train_features)
    predictions_valid_rfc = model_rfc.predict(valid_features)
    
    score_train_rfc = accuracy_score(predictions_train_rfc, train_target)
    score_valid_rfc = accuracy_score(predictions_valid_rfc, valid_target)
    
    print(f'Accuracy при max_depth = {est}')
    print(f'Обучающая выборка: {score_train_rfc}')
    print(f'Валидационная выборка: {score_valid_rfc}')
    print()
    
    if score_valid_rfc > best_result_dtc:
        best_result_rfc = score_valid_rfc
        best_est_rfc = est
        
    
print(f'Лучший показатель Accuracy: {best_result_rfc.round(4)} при n_estimators = {best_est_rfc}')

Accuracy при max_depth = 3
Обучающая выборка: 0.9499066583696328
Валидационная выборка: 0.9527363184079602

Accuracy при max_depth = 4
Обучающая выборка: 0.9467952706907281
Валидационная выборка: 0.9502487562189055

Accuracy при max_depth = 5
Обучающая выборка: 0.9738643434971997
Валидационная выборка: 0.9751243781094527

Accuracy при max_depth = 6
Обучающая выборка: 0.970130678282514
Валидационная выборка: 0.972636815920398

Accuracy при max_depth = 7
Обучающая выборка: 0.9803982576228998
Валидационная выборка: 0.9900497512437811

Accuracy при max_depth = 8
Обучающая выборка: 0.9757311761045426
Валидационная выборка: 0.9776119402985075

Accuracy при max_depth = 9
Обучающая выборка: 0.9835096453018046
Валидационная выборка: 0.9850746268656716

Accuracy при max_depth = 10
Обучающая выборка: 0.9813316739265713
Валидационная выборка: 0.9850746268656716

Лучший показатель Accuracy: 0.9851 при n_estimators = 10


### Исследуем модель Logistic Regression

In [59]:
# Обучение и предсказание модели

#model_lr = LogisticRegression(random_state = 12345)
model_lr = LogisticRegression(random_state = 12345, solver = 'lbfgs', max_iter = 1000)
model_lr.fit(train_features, train_target)

predictions_train_lr = model_lr.predict(train_features)
predictions_valid_lr = model_lr.predict(valid_features)

score_train_lr = accuracy_score(train_target, predictions_train_lr)
score_valid_lr = accuracy_score(valid_target, predictions_valid_lr)

print(f'Обучающая выборка: {score_train_lr.round(4)}')
print(f'Валидационная выборка: {score_valid_lr.round(4)}')

Обучающая выборка: 0.748
Валидационная выборка: 0.7736


### *Вывод по разделу исследования моделей*

Результаты наилучших показателей Accuracy среди моделей распределились следующим образом:  
- Accuracy `Decision Tree Classifier: ` 0.8706, при depth = 10
- Accuracy `Random Forest Classifier: ` 0.9851, при n_estimators = 10
- Accuracy `Logistic Regression: ` 0.7736, при solver = 'lbfgs'и max_iter = 1000

## Проверка моделей на тестовой выборке

In [60]:
# Подготовка данных тестовой выборки для проверки моделей

final_test_features = final_test_data.drop(['is_ultra'], axis = 1)
final_test_target = final_test_data['is_ultra']

Наиболее точной себя показала модель случайного леса - Random Forest Classifier, поэтому проверим ее на тестовой выборке

### Проверка Decision Tree Classifier на тестовой выборке

In [61]:

model_dtc = DecisionTreeClassifier(random_state = 12345, max_depth = 10)
model_dtc.fit(final_test_features, final_test_target)

predictions_test_dtc = model_dtc.predict(final_test_features)
score_test_dtc = accuracy_score(final_test_target, predictions_test_dtc)

print(f'Accuracy Decision Tree Classifier: {score_test_dtc.round(4)}')

Accuracy Decision Tree Classifier: 0.9204


### Проверка Random Forest Classifier на тестовой выборке

In [62]:

#model_rfc = RandomForestClassifier(random_state = 12345, n_estimators = est)
model_rfc = RandomForestClassifier(random_state = 12345, n_estimators = 10)
model_rfc.fit(final_test_features, final_test_target)

predictions_test_rfc = model_rfc.predict(final_test_features)
score_test_rfc = accuracy_score(predictions_test_rfc, final_test_target)

print(f'Accuracy Random Forest Classifier: {score_test_rfc.round(4)}')

Accuracy Random Forest Classifier: 0.9652


### Проверка Logistic Regression на тестовой выборке

In [63]:

#model_lr = LogisticRegression(random_state = 12345)
model_lr = LogisticRegression(random_state = 12345, solver = 'lbfgs', max_iter = 1000)
model_lr.fit(final_test_features, final_test_target)

predictions_test_lr = model_lr.predict(final_test_features)
score_test_lr = accuracy_score(predictions_test_lr, final_test_target)

print(f'Accuracy Logistic Regression: {score_test_lr.round(4)}')

Accuracy Logistic Regression: 0.704


### *Вывод по разделу проверки моделей на тестовой выборке*

- Лучший показатель точности на тестовой выборке оказался у модели `Random Forest Classifier`, и составил `0.9652`  
- Точность `Decision Tree Classifier` на тестовой выборке составила - `0.9204`  
- Модель `Logistic Regression` не зарекомендовала себя ни в процессе обучения, ни на тестовой выборке. Лучшая точность модели, не достигнув минимального допустимого порога, составила `0.7736` и `0.704` соответственно

## Проверка модели на адекватность

Проверку на вменяемость модели можно проверить, выполнив простое сравнение точности модели со случайным распределением:

In [64]:
# Проверка модели Random Forest Classifier на адекватность

if score_test_rfc > 0.5:
    print(f'Точность модели больше 0.5 и составляет {score_test_rfc.round(4)}. Модель адекватна')
else:
    print(f'Точность модели меньше 0.5 и составляет {score_test_rfc.round(4)}. Модель неадекватна')
    

Точность модели больше 0.5 и составляет 0.9652. Модель адекватна


## *Общий вывод*

- Для анализа были представлены мобильного оператора о поведении пользователей за месяц: количество звонков, суммарная длительность звонков в минутах, количество sms-сообщений, израсходованный интернет-трафик в Мб, каким тарифом пользовался в течение месяца.  
- При предобработке данных был изменен тип столбцоа calls и messages на int для удобства исследования  
- Были исследованы три модели `Decision Tree Classifier`, `Random Forest Classifier` и `Logistic Regression`  
- Лучшую точность `0.9652` показала модель `Random Forest Classifier`с гиперпараметрами `random_state = 12345` и `n_estimators = 10`  
- Точность моделей `Decision Tree Classifier` и `Logistic Regression` составили `0.9204` и `0.704` соответственно.
- Адекватность модели была проверена путем сравнения с точности модели со случайным распределением (`0.5`). Модель `Random Forest Classifier` адекватна  